src/main/api/catalog.ts

```diff
@@ -138,104 +137,215 @@ export class CatalogApi implements ICatalogApi {
-                    const r2PublicationJson = JSON.parse(r2PublicationStr);
-                    const r2Publication = TaJsonDeserialize<R2Publication>(r2PublicationJson, R2Publication);
-                    return !isAudiobookFn(view) && !isDivinaFn(view) && !isPdfFn(r2Publication);
-                },
-            );
+            return r2Publication;
+        };
+
+        const lastAddedPublicationsDocumentRaw = await this.getLastAddedPublicationDocument();
+        const lastReadingPubArray = this.getLastReadingPublicationId();
+
+        const lastAddedPublicationsDocument =
+            lastAddedPublicationsDocumentRaw.filter(({ identifier }) => !lastReadingPubArray.includes(identifier));
+        const lastReadedPublicationDocument =
+            lastReadingPubArray.map(
+                (identifier) => lastAddedPublicationsDocumentRaw.find((v) => v.identifier === identifier));
+
+        const lastAddedPublicationsView =
+            lastAddedPublicationsDocument.map((doc) => {
+                try {
+                    return this.publicationViewConverter.convertDocumentToView(doc);
+                } catch (e) {
+                    debug("lastadded publication view converter", e);
+                    errorDeletePub(doc);
+                }
 
+                return undefined;
+            }).filter((v) => !!v);
+        const lastReadedPublicationsView =
+            lastReadedPublicationDocument.map((doc) => {
+                try {
+                    return this.publicationViewConverter.convertDocumentToView(doc);
+                } catch (e) {
+                    debug("lastreaded publication view converter", e);
+                    errorDeletePub(doc);
+                }
+
+                return undefined;
+            }).filter((v) => !!v);
+
+        // {
+
+        //     // debug(lastAddedPublicationsView[0]);
+        //     const { identifier, title } = lastReadedPublicationsView[0];
+        //     debug(identifier, title);
+        // }
+
+        const audio = {
+            readed: lastReadedPublicationsView.filter(isAudiobookFn),
+            added: lastAddedPublicationsView.filter(isAudiobookFn),
         };
 
-        const lastAddedPublicationsDocument = await this.getLastAddedPublicationDocument();
+        const divina = {
+            readed: lastReadedPublicationsView.filter(isDivinaFn),
+            added: lastAddedPublicationsView.filter(isDivinaFn),
+        };
 
-        const lastAddedPublicationsDocumentArray = lastAddedPublicationsDocument.map(
-            (doc) => this.getLastReadingPublicationId().includes(doc.identifier)
-                ? [doc, undefined]
-                : [undefined, doc],
-        );
+        const pdf = {
+            readed: lastReadedPublicationsView.filter(
+                (view: PublicationView) => {
+                    const r2Publication = viewToR2Pub(view);
```

Comment:
Pierre, as you remarked before, each call to the `viewToR2Pub()` function is computationally expensive. As it is only necessary for `isPdfFn()`, is there a way we could cache some kind of `isPDF` boolean?

Reply:
you're right , need to memoize ..
perhaps. this: https://github.com/ramda/ramda/tree/v0.27.0/source/memoizeWith.js

Reply:
https://github.com/edrlab/thorium-reader/pull/1292/commits/6c4b1e1b28eb72c819c081f96e7e9d31ecaf13f1

Reply:
Good idea. `R.identity` for `view` will not work, right? (because the object pointer is different at each invocation, I think)

Reply:
Ah, I have just seen your commit with `PublicationView.identifier` ... great :)